8331947: Preview creates checkbox for JEP-less preview feature

hns opened the pull request:
Please review a simple patch to exclude preview visitor classes meant to support future preview features from the Preview API page. 

The test adds an sample element annotated with the new `PreviewFeature.Feature.LANGUAGE_MODEL` constant (which does not have a `@JEP` annotation) to make sure it is not listed in the Preview API page. The test itself does not have to be modified, as it would fail without the change in `PreviewAPIListBuilder.java`.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331947](https://bugs.openjdk.org/browse/JDK-8331947): Preview creates checkbox for JEP-less preview feature (**Bug** - P4)


### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - Author)
 * [Pavel Rappo](https://openjdk.org/census#prappo) (@pavelrappo - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19344/head:pull/19344` \
`$ git checkout pull/19344`

Update a local copy of the PR: \
`$ git checkout pull/19344` \
`$ git pull https://git.openjdk.org/jdk.git pull/19344/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19344`

View PR using the GUI difftool: \
`$ git pr show -t 19344`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19344.diff">https://git.openjdk.org/jdk/pull/19344.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19344#issuecomment-2124210880)

liach commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/PreviewAPIListBuilder.java:
```diff
@@ -86,7 +86,12 @@ protected void handleElement(Element e) {
-        elementJeps.put(e, jep);
+        if (jep.number == 0) {
+            // Remove preview support features without a valid JEP
+            jeps.remove(feature);
```
This feels like we are adding and removing a dummy jep to and from the map every time we encounter the language model feature. Can we just bind a dummy jep to that feature to avoid repeated processing, and only filter at the getter methods?

hns replied:
I agree it's not ideal. I think the best thing is to split up the insertion into separate steps: First check if the key is contained in the map, retrieve the JEP info if it isn't, and only then decide whether to create a JEP record.

But I found another problem with this PR, so there may be some further changes coming.

hns commented:
Note: this patch only removes the checkbox labeled "0: (Preview)" from the top of the Preview API page, but the preview visitor classes themselves are still contained in the page. I'm not sure they need to be, so this PR may see further updates.

hns commented:
I updated this PR to exclude elements of preview features that don't have a JEP from the Preview API page. 

In order to do this, I changed the way `SummaryAPIListBuilder` determines which API elements to include from using a `Predicate` passed to the constructor to using a new abstract `belongsToSummary(Element)` method. The reason for this is that the predicate had to be passed to the builder's super-constructor, which made it impossible to use an instance method of the current builder. However, since determining whether a preview API element has a JEP is somewhat involved (retrieving annotation elements of an enum constant used as annotation element), we only want to do this once and store the results. Being able to use an ordinary instance method was the simplest solution to this problem.

I also changed the workaround method that retrieves the JEP info to return a `Map<String, Object>` instead of the the raw `javax.lang.model.element` types, which makes it easier to use.

hns commented:
Thanks @liach, I updated the PR.

pavelrappo commented on test/langtools/jdk/javadoc/doclet/testPreview/TestPreview.java:
```diff
@@ -156,6 +157,9 @@ public void testPreviewAPIJavadoc() {
                     <li><a href="Core.html" class="current-selection">Core</a></li>
                     </ol>""");
+
+        // 8331947: Support preview features without JEP should not be included in Preview API page
```
Trivial: here it is phrased as "Support preview features", while in the actual test source file it is "Preview support feature".

hns replied:
I agree the terminology in comments is somewhat inconsistent. I don't think there is an "official" term for these preview features. [In the CSR](https://bugs.openjdk.org/browse/JDK-8329634) they are summarized as follows: "Add a set of persistent preview visitors to support new language changes in development." 

I think the essential property for our purpose is the "without JEP". I added the "support" adjective to make their purpose clearer, but I don't think it's essential.

pavelrappo commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/PreviewAPIListBuilder.java:
```diff
@@ -60,40 +60,51 @@ public int compareTo(JEP o) {
+            elementJeps.put(element, jep);
+            return true;
+        }
+        // Preview features without JEP are not included.
```
Trivial: we don't need to mention "support" near "preview" and "features", right? Just checking.

pavelrappo reviewed (approved):
Approved with fuzz.

hns commented:
/integrate
